### PR TITLE
DPL-908 PBMC Isolations

### PR DESCRIPTION
**Pull request by yoldas**
Closes #

#### Changes proposed in this pull request

- Labware creator to pool wells with the same samples to into destination well
- Group source wells into pairs or configured number of source wells
- Remaining or odd number of source cells are transferred to its own well
- Ignore failed source wells
- Driver file

#### Instructions for Reviewers

_[All PRs] - Confirm PR template filled_  
_[Feature Branches] - Review code_  
_[Production Merges to `main`]_  
 &nbsp; &nbsp; \- _Check story numbers included_  
 &nbsp; &nbsp; \- _Check for debug code_  
 &nbsp; &nbsp; \- _Check version_

### CODE REVIEW

File: app/models/labware_creators/pooled_wells_by_sample.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+require_dependency 'labware_creators/base'
+
+module LabwareCreators
+  # This labware creator pools PBMC isolations with the same samples in pairs
+  # (or configured number of source wells) per destination before cell counting
+  # to reduce the number runs on Celleca (cell counting). The robot transfer is
+  # done from LRC Blood Bank to LRC PBMC Bank plate. Only the wells with passed
+  # state will be transferred to the destination plate. The destination wells
+  # are compressed to top left by column on the plate.
+  class PooledWellsBySample < Base
```
Comment (andrewsparkes) [inline app/models/labware_creators/pooled_wells_by_sample.rb:12]: I think we need a more descriptive name for this creator that suggests the grouping in pairs / a number.
Maybe PooledWellsBySampleInGroups ? PooledWellsBySampleGrouped ? PooledWellsBySampleInSets ? PooledWellsBySampleInLots ?

File: app/models/labware_creators/pooled_wells_by_sample.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      @well_filter ||= WellFilter.new(creator: self)
+    end
+
+    def filters=(filter_parameters)
+      well_filter.assign_attributes(filter_parameters)
+    end
+
+    # Parent plate using SS v2 API
+    def source_plate
+      @source_plate ||= Sequencescape::Api::V2::Plate.find_by(uuid: parent.uuid)
```
Comment (andrewsparkes) [inline app/models/labware_creators/pooled_wells_by_sample.rb:33]: This is bypassing the well filter code. I'm uncertain if that's risky.
Well filter pulls the list of wells it checks from the labware creator `labware_wells` method. You could do what I did and filter the wells there. e.g. I did this to get rid of failed wells:
```
def labware_wells
    parent.wells_in_columns.filter_map { |well| well unless well.failed? }
end
```
I do not know what is the right solution.

Comment (andrewsparkes) [reply app/models/labware_creators/pooled_wells_by_sample.rb:33]: I also ordered them by column there.

File: app/models/labware_creators/pooled_wells_by_sample.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+      plate.wells.detect { |well| well.location == well_location }
+    end
+
+    # Attributes for transfer request from source_well to dest_plate
+    def request_hash(source_well, dest_plate)
+      dest_location = transfer_hash[source_well.location][:dest_locn]
+      {
+        'source_asset' => source_well.uuid,
+        'target_asset' => get_well_for_plate_location(dest_plate, dest_location)&.uuid,
+        # 'submission' => submission,
```
Comment (andrewsparkes) [inline app/models/labware_creators/pooled_wells_by_sample.rb:82]: The `submission_id` comes in through the `well_filter.filtered` method as `[well, additional_params]` where additional_params is `{.'submission_id': 1 }`, it extracts the submission id from the wells. It is normal to have it in the transfers.

**Review by andrewsparkes — CHANGES_REQUESTED** [review]
I think get the submission_id into the transfers is the main thing that needs changing.
I am unsure what to do about the WellFilter.

We are going to have merge conflicts with the other driver file story.

File: app/models/labware_creators/pooled_wells_by_sample.rb (hunk 1 above)
Comment (yoldas) [reply app/models/labware_creators/pooled_wells_by_sample.rb:12]: More specific and descriptive name should be given. PooledWellsBySampleInGroups can be used unless any other can be found.

File: config/purposes/scrna_core_cell_extraction.wip.yml [hunk 4]
```diff
@@ -52,7 +52,12 @@ LRC Blood Bank:
 # Output plate from Hamilton PBMC isolation protocol / input plate for the cell counting and cell banking protocols
 LRC PBMC Bank:
   :asset_type: plate
+  # Labware creator for pooling isolations by sample before cell counting
+  :creator_class: LabwareCreators::PooledWellsBySampleInGroups
   :label_template: plate_cellaca_qc # creates QC1 up to QC4 barcodes
+  # number of source wells to pool into a single destination well by sample from LRC Bank to LRC PBMC Bank
+  # e.g. 6 copies of sample in 6 wells, '2' in the config, pool to make 3 child wells
+  :number_of_source_wells: 2
```
Comment (andrewsparkes) [inline config/purposes/scrna_core_cell_extraction.wip.yml:60]: if `number_of_source_wells` is specific to the labware creator you could use the args syntax like on lines 97-102 below so it's kept together

**Review by andrewsparkes — APPROVED** [review]

**Pull request merged**